test/integration/evictions/evictions_test.go

```diff
@@ -38,12 +38,14 @@ import (
 	"k8s.io/apimachinery/pkg/types"
```

Comment:
I thought of one more scenario where the update and delete behavior could differ, which was if the pod we were trying to manipulate was already gone from etcd. The Delete function has some compensation for that which Update does not (since if you are asking to delete an object, and it doesn't exist, Delete could plausibly return success)

To prove that addConditionAndDeletePod returns identical responses with/without the feature gate enabled, I made this test:

```go
func TestAddConditionAndDelete(t *testing.T) {
	cases := []struct {
		name              string
		initialPod        bool
		makeDeleteOptions func(*api.Pod) *metav1.DeleteOptions
		expectErr         string
	}{
		{
			name:              "simple",
			initialPod:        true,
			makeDeleteOptions: func(pod *api.Pod) *metav1.DeleteOptions { return &metav1.DeleteOptions{} },
		},
		{
			name:              "missing",
			initialPod:        false,
			makeDeleteOptions: func(pod *api.Pod) *metav1.DeleteOptions { return &metav1.DeleteOptions{} },
			expectErr:         "not found",
		},
		{
			name:       "valid uid",
			initialPod: true,
			makeDeleteOptions: func(pod *api.Pod) *metav1.DeleteOptions {
				return &metav1.DeleteOptions{Preconditions: &metav1.Preconditions{UID: &pod.UID}}
			},
		},
		{
			name:       "invalid uid",
			initialPod: true,
			makeDeleteOptions: func(pod *api.Pod) *metav1.DeleteOptions {
				badUID := pod.UID + "1"
				return &metav1.DeleteOptions{Preconditions: &metav1.Preconditions{UID: &badUID}}
			},
			expectErr: "The object might have been deleted and then recreated",
		},
		{
			name:       "valid resourceVersion",
			initialPod: true,
			makeDeleteOptions: func(pod *api.Pod) *metav1.DeleteOptions {
				return &metav1.DeleteOptions{Preconditions: &metav1.Preconditions{ResourceVersion: &pod.ResourceVersion}}
			},
		},
		{
			name:       "invalid resourceVersion",
			initialPod: true,
			makeDeleteOptions: func(pod *api.Pod) *metav1.DeleteOptions {
				badRV := pod.ResourceVersion + "1"
				return &metav1.DeleteOptions{Preconditions: &metav1.Preconditions{ResourceVersion: &badRV}}
			},
			expectErr: "The object might have been modified",
		},
	}

	testContext := genericapirequest.WithNamespace(genericapirequest.NewContext(), metav1.NamespaceDefault)

	storage, _, _, server := newStorage(t)
	defer server.Terminate(t)
	defer storage.Store.DestroyFunc()

	client := fake.NewSimpleClientset()
	evictionRest := newEvictionStorage(storage.Store, client.PolicyV1())

	for _, tc := range cases {
		for _, conditionsEnabled := range []bool{true, false} {
			name := fmt.Sprintf("%s_conditions=%v", tc.name, conditionsEnabled)
			t.Run(name, func(t *testing.T) {
				defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.PodDisruptionConditions, conditionsEnabled)()
				var deleteOptions *metav1.DeleteOptions
				if tc.initialPod {
					newPod := validNewPod()
					createdObj, err := storage.Create(testContext, newPod, rest.ValidateAllObjectFunc, &metav1.CreateOptions{})
					if err != nil {
						t.Fatal(err)
					}
					t.Cleanup(func() {
						zero := int64(0)
						storage.Delete(testContext, newPod.Name, rest.ValidateAllObjectFunc, &metav1.DeleteOptions{GracePeriodSeconds: &zero})
					})
					deleteOptions = tc.makeDeleteOptions(createdObj.(*api.Pod))
				} else {
					deleteOptions = tc.makeDeleteOptions(nil)
				}
				if deleteOptions == nil {
					deleteOptions = &metav1.DeleteOptions{}
				}

				err := addConditionAndDeletePod(evictionRest, testContext, "foo", rest.ValidateAllObjectFunc, deleteOptions)
				if err == nil {
					if tc.expectErr != "" {
						t.Fatalf("expected err containing %q, got none", tc.expectErr)
					}
					return
				}
				if tc.expectErr == "" {
					t.Fatalf("unexpected err: %v", err)
				}
				if !strings.Contains(err.Error(), tc.expectErr) {
					t.Fatalf("expected err containing %q, got %v", tc.expectErr, err)
				}
			})
		}
	}
}
```

That shows we're consistent in the missing pod case as well, so no changes to `pkg/registry/core/pod/storage/eviction.go` are needed, but if you want to add that test, I think that's a good way to exercise that function with/without the feature gate on and make sure the same inputs produce the same outputs.

Reply:
I have included the test, thanks! Good to be thorough and compare the possible scenarios.